core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java

```diff
@@ -0,0 +1,372 @@
+
+        // Calculate minimum and maximum size for candidate strings
+        // This is used for short-circuiting the match if the size of
+        // the input is outside those bounds
+        int minSize = 0;
+        int maxSize = 0;
+        boolean unbounded = false;
+        for (Pattern expression : optimized) {
+            if (expression instanceof Literal literal) {
+                int length = literal.value().getBytes(UTF_8).length;
+                minSize += length;
+                maxSize += length;
+            }
+            else if (expression instanceof Any any) {
+                int length = any.min();
+                minSize += length;
+                maxSize += length * 4; // at most 4 bytes for a single UTF-8 codepoint
+
+                unbounded = unbounded || any.unbounded();
+            }
+            else {
+                throw new UnsupportedOperationException("Not supported: " + expression.getClass().getName());
+            }
+        }
+
+        // Calculate exact match prefix and suffix
+        // If the pattern starts and ends with a literal, we can perform a quick
+        // exact match to short-circuit DFA evaluation
+        byte[] prefix = new byte[0];
+        byte[] suffix = new byte[0];
+        List<Pattern> middle = new ArrayList<>();
+        for (int i = 0; i < optimized.size(); i++) {
+            Pattern expression = optimized.get(i);
+
+            if (i == 0) {
+                if (expression instanceof Literal literal) {
+                    prefix = literal.value().getBytes(UTF_8);
+                    continue;
+                }
+            }
+            else if (i == optimized.size() - 1) {
+                if (expression instanceof Literal literal) {
+                    suffix = literal.value().getBytes(UTF_8);
+                    continue;
+                }
+            }
+
+            middle.add(expression);
+        }
+
+        // If the pattern (after excluding constant prefix/suffixes) ends with an unbounded match (i.e., %)
+        // we can perform a non-exact match and end as soon as the DFA reaches an accept state -- there
+        // is no need to consume the remaining input
+        // This section determines whether the pattern is a candidate for non-exact match.
+        boolean exact = true; // whether to match to the end of the input
+        if (!middle.isEmpty()) {
+            // guaranteed to be Any because any Literal would've been turned into a suffix above
+            Any last = (Any) middle.get(middle.size() - 1);
+            if (last.unbounded()) {
+                exact = false;
+
+                // Since the matcher will stop early, no need for an unbounded matcher (it produces a simpler DFA)
+                middle.set(middle.size() - 1, new Any(last.min(), false));
+            }
+        }
+
+        NFA nfa = makeNfa(middle);
+        DFA dfa = nfa.toDFA();
+
+        int[] transitions = new int[dfa.states().size() * 256];
```

Comment:
It's not significant in general that it's worth tracking. The memory utilized will be proportional to the length of the pattern * 256, so it will be just a few KB at most in general.